src/instructlab/model/download.py

```diff
@@ -12,6 +16,236 @@
+
+        try:
+            with open(
+                f"{oci_model_path}/blobs/sha256/{index_hash}",
+                mode="r",
+                encoding="UTF-8",
+            ) as f:
+                index = json.load(f)
+        except FileNotFoundError as exc:
+            raise ValueError(f"file not found: {oci_model_path}/index.json") from exc
+        except json.JSONDecodeError as exc:
+            raise ValueError(
+                f"could not read JSON file: {oci_model_path}/index.json"
+            ) from exc
+        except Exception as exc:
+            raise ValueError("unexpected error occurred: {e}") from exc
+
+        title_ref = "org.opencontainers.image.title"
+        oci_model_file_map = {}
+
+        for layer in index["layers"]:
+            match = re.search("sha256:(.*)", layer["digest"])
+
+            if match:
+                blob_name = match.group(1)
+                oci_model_file_map[blob_name] = layer["annotations"][title_ref]
+
+        return oci_model_file_map
+
+    def download(self):
+        click.echo(
+            f"Downloading model from OCI registry: {self.repository}@{self.release} to {self.download_dest}..."
+        )
+
+        os.makedirs(self.download_dest, exist_ok=True)
+        model_name = self.repository.split("/")[-1]
+        oci_dir = f"{DEFAULTS.OCI_DIR}/{model_name}"
+        os.makedirs(oci_dir, exist_ok=True)
+
+        if not is_tool_present("skopeo"):
+            click.secho(
+                "skopeo not detected, but required to perform downloads from OCI registries. Exiting",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+        try:
+            subprocess.run(
+                [
+                    "skopeo",
+                    "copy",
+                    f"{self.repository}:{self.release}",
+                    f"oci:{oci_dir}",
+                ],
+                check=True,
+            )
+        except Exception as e:
+            raise ValueError(
+                "CalledProcessError: command exited with non-zero code"
+            ) from e
+
+        file_map = self._build_oci_model_file_map(oci_dir)
+
+        for _, _, files in os.walk(f"{oci_dir}/blobs/sha256/"):
+            for name in files:
+                if name not in file_map:
+                    continue
+                dest = file_map[name]
+                if not os.path.exists(os.path.join(self.download_dest, model_name)):
+                    os.makedirs(
+                        os.path.join(self.download_dest, model_name), exist_ok=True
+                    )
+                # unlink any existing version of the file
+                if os.path.exists(os.path.join(self.download_dest, model_name, dest)):
+                    os.unlink(os.path.join(self.download_dest, model_name, dest))
+
+                # create hard link to files in cache, to avoid redownloading if model has been downloaded before
```

Comment:
Why a hard link and not a symlink?

Reply:
I suppose there isn't a strong reason for that, maybe a symlink would work as well
the only thing is maybe in case we lose either the cache or the models_dir for whatever reason, we wouldn't have to redownlaod the models with a hard link but we would if it was a symlink

Reply:
hardlinks won't work across file systems. Is it always the case that the cache and model_dir are on the same FS? I believe @tiran mentioned that he hosts the data dir on NFS while keeping cache local. I think it won't work then.

Reply:
I think it makes sense to switch this to fully qualified symlinks. Relative can sometimes cause issues with container volume mounting.

Reply:
switched to symlinks